File: sdks/typescript/tests/multi-database.test.ts
import { describe, it, expect, beforeAll, afterAll } from 'vitest';
import { NexusClient } from '../src';

describe('Multi-Database Support', () => {
  let client: NexusClient;

  beforeAll(() => {
    client = new NexusClient({
      baseUrl: 'http://localhost:15474',
    });
  });

  afterAll(async () => {
    // Clean up any test databases that might still exist
    try {
      const databases = await client.listDatabases();
      const testDbs = databases.databases.filter((db: string) =>
        db.startsWith('test_')
      );

      // Switch to default database first
      if (testDbs.length > 0) {
        await client.switchDatabase(databases.defaultDatabase);
      }

      // Drop test databases
      for (const db of testDbs) {
        try {
          await client.dropDatabase(db);
        } catch (e) {
          // Ignore errors during cleanup
        }
      }
    } catch (e) {
      // Ignore errors during cleanup
    }
  });

  it('should list databases', async () => {
    const databases = await client.listDatabases();

    expect(databases.databases).toBeDefined();
    expect(Array.isArray(databases.databases)).toBe(true);
    expect(databases.databases.length).toBeGreaterThan(0);
    expect(databases.defaultDatabase).toBeDefined();
    expect(databases.databases).toContain(databases.defaultDatabase);
  });

  it('should create and drop a database', async () => {
    const dbName = 'test_temp_db';

    // Create database
    const createResult = await client.createDatabase(dbName);
    expect(createResult.success).toBe(true);
    expect(createResult.name).toBe(dbName);

    // Verify it exists
    let databases = await client.listDatabases();
    expect(databases.databases).toContain(dbName);

    // Drop database
    const dropResult = await client.dropDatabase(dbName);
    expect(dropResult.success).toBe(true);

    // Verify it's gone
    databases = await client.listDatabases();
    expect(databases.databases).not.toContain(dbName);
  });

  it('should switch between databases', async () => {
    const dbName = 'test_switch_db';

    try {
      // Create a test database
      await client.createDatabase(dbName);

      // Get initial database
      const initialDb = await client.getCurrentDatabase();

      // Switch to test database
      const switchResult = await client.switchDatabase(dbName);
      expect(switchResult.success).toBe(true);

      // Verify we're in the new database
      let currentDb = await client.getCurrentDatabase();
      expect(currentDb).toBe(dbName);

      // Switch back
      const switchBack = await client.switchDatabase(initialDb);
      expect(switchBack.success).toBe(true);

      // Verify we're back
      currentDb = await client.getCurrentDatabase();
      expect(currentDb).toBe(initialDb);
    } finally {
      // Clean up
      const databases = await client.listDatabases();
      await client.switchDatabase(databases.defaultDatabase);
      await client.dropDatabase(dbName);
    }
  });

  it('should get database information', async () => {
    const dbName = 'test_info_db';

    try {
      // Create a test database
      await client.createDatabase(dbName);

      // Get database info
      const dbInfo = await client.getDatabase(dbName);
      expect(dbInfo.name).toBe(dbName);
      expect(dbInfo.path).toBeDefined();
      expect(typeof dbInfo.nodeCount).toBe('number');
      expect(typeof dbInfo.relationshipCount).toBe('number');
      expect(typeof dbInfo.storageSize).toBe('number');
      expect(dbInfo.nodeCount).toBeGreaterThanOrEqual(0);
      expect(dbInfo.relationshipCount).toBeGreaterThanOrEqual(0);
      expect(dbInfo.storageSize).toBeGreaterThanOrEqual(0);
    } finally {
      // Clean up
      await client.dropDatabase(dbName);
    }
  });

  it('should isolate data between databases', async () => {
    const db1Name = 'test_isolation_db1';
    const db2Name = 'test_isolation_db2';

    try {
      // Create two test databases
      await client.createDatabase(db1Name);
      await client.createDatabase(db2Name);

      // Switch to db1 and create a node
      await client.switchDatabase(db1Name);
      let result = await client.executeCypher(
        'CREATE (n:TestNode {name: $name}) RETURN n',
        { name: 'DB1 Node' }
      );
      expect(result.rows.length).toBe(1);

      // Verify node exists in db1
      result = await client.executeCypher(
        'MATCH (n:TestNode) RETURN count(n) AS count',
        {}
      );
      expect(result.rows[0].count).toBe(1);

      // Switch to db2
      await client.switchDatabase(db2Name);

      // Verify node does NOT exist in db2 (isolation)
      result = await client.executeCypher(
        'MATCH (n:TestNode) RETURN count(n) AS count',
        {}
      );
      expect(result.rows[0].count).toBe(0);

      // Create a different node in db2
      result = await client.executeCypher(
        'CREATE (n:TestNode {name: $name}) RETURN n',
        { name: 'DB2 Node' }
      );
      expect(result.rows.length).toBe(1);

      // Verify only one node in db2
      result = await client.executeCypher(
        'MATCH (n:TestNode) RETURN count(n) AS count',
        {}
      );
      expect(result.rows[0].count).toBe(1);

      // Switch back to db1
      await client.switchDatabase(db1Name);

      // Verify still only one node in db1
      result = await client.executeCypher(
        'MATCH (n:TestNode) RETURN count(n) AS count',
        {}
      );
      expect(result.rows[0].count).toBe(1);
    } finally {
      // Clean up
      const databases = await client.listDatabases();
      await client.switchDatabase(databases.defaultDatabase);
      await client.dropDatabase(db1Name);
      await client.dropDatabase(db2Name);
    }
  });

  it('should connect to a specific database using constructor', async () => {
    const dbName = 'test_param_db';

    try {
      // Create a test database
      await client.createDatabase(dbName);

      // Create a new client connected to the specific database
      const dbClient = new NexusClient({
        baseUrl: 'http://localhost:15474',
        database: dbName,
      });

      // Verify we're connected to the right database
      const currentDb = await dbClient.getCurrentDatabase();
      expect(currentDb).toBe(dbName);
    } finally {
      // Clean up
      await client.dropDatabase(dbName);
    }
  });

  it('should not allow dropping the current database', async () => {
    const dbName = 'test_no_drop_db';

    try {
      // Create a test database
      await client.createDatabase(dbName);

      // Switch to the database
      await client.switchDatabase(dbName);

      // Try to drop it while it's active - should fail
      await expect(client.dropDatabase(dbName)).rejects.toThrow();

      // Switch to a different database
      const databases = await client.listDatabases();
      await client.switchDatabase(databases.defaultDatabase);

      // Now we should be able to drop it
      const dropResult = await client.dropDatabase(dbName);
      expect(dropResult.success).toBe(true);
    } catch (error) {
      // Clean up even if test fails
      const databases = await client.listDatabases();
      await client.switchDatabase(databases.defaultDatabase);

      try {
        await client.dropDatabase(dbName);
      } catch (e) {
        // Ignore cleanup errors
      }

      throw error;
    }
  });

  it('should not allow dropping the default database', async () => {
    // Get default database
    const databases = await client.listDatabases();
    const defaultDb = databases.defaultDatabase;

    // Try to drop it - should fail
    await expect(client.dropDatabase(defaultDb)).rejects.toThrow();
  });
});

File: sdks/rust/tests/multi_database_test.rs
//! Tests for multi-database support

use nexus_sdk::{NexusClient, NexusError, Value};
use std::collections::HashMap;

#[tokio::test]
async fn test_list_databases() -> Result<(), NexusError> {
    let client = NexusClient::new("http://localhost:15474")?;

    let databases = client.list_databases().await?;

    assert!(!databases.databases.is_empty());
    assert!(!databases.default_database.is_empty());
    // Check that default database exists in the list
    assert!(
        databases
            .databases
            .iter()
            .any(|db| db.name == databases.default_database)
    );

    Ok(())
}

#[tokio::test]
async fn test_create_and_drop_database() -> Result<(), NexusError> {
    let client = NexusClient::new("http://localhost:15474")?;

    let db_name = "test_temp_db";

    // Create database
    let create_result = client.create_database(db_name).await?;
    assert!(create_result.success);
    assert_eq!(create_result.name, db_name);

    // Verify it exists
    let mut databases = client.list_databases().await?;
    assert!(databases.databases.iter().any(|db| db.name == db_name));

    // Drop database
    let drop_result = client.drop_database(db_name).await?;
    assert!(drop_result.success);

    // Verify it's gone
    databases = client.list_databases().await?;
    assert!(!databases.databases.iter().any(|db| db.name == db_name));

    Ok(())
}

#[tokio::test]
async fn test_switch_database() -> Result<(), NexusError> {
    let client = NexusClient::new("http://localhost:15474")?;

    let db_name = "test_switch_db";

    // Create a test database
    client.create_database(db_name).await?;

    // Get initial database
    let initial_db = client.get_current_database().await?;

    // Switch to test database
    let switch_result = client.switch_database(db_name).await?;
    assert!(switch_result.success);

    // Verify we're in the new database
    let mut current_db = client.get_current_database().await?;
    assert_eq!(current_db, db_name);

    // Switch back
    let switch_back = client.switch_database(&initial_db).await?;
    assert!(switch_back.success);

    // Verify we're back
    current_db = client.get_current_database().await?;
    assert_eq!(current_db, initial_db);

    // Clean up
    client.drop_database(db_name).await?;

    Ok(())
}

#[tokio::test]
async fn test_get_database_info() -> Result<(), NexusError> {
    let client = NexusClient::new("http://localhost:15474")?;

    let db_name = "test_info_db";

    // Create a test database
    client.create_database(db_name).await?;

    // Get database info
    let db_info = client.get_database(db_name).await?;
    assert_eq!(db_info.name, db_name);
    assert!(!db_info.path.is_empty());
    // node_count, relationship_count, and storage_size are u64, so always >= 0
    // Just verify they exist (the struct was successfully deserialized)
    let _ = db_info.node_count;
    let _ = db_info.relationship_count;
    let _ = db_info.storage_size;

    // Clean up
    client.drop_database(db_name).await?;

    Ok(())
}

#[tokio::test]
async fn test_data_isolation() -> Result<(), NexusError> {
    let client = NexusClient::new("http://localhost:15474")?;

    let db1_name = "test_isolation_db1";
    let db2_name = "test_isolation_db2";

    // Create two test databases
    client.create_database(db1_name).await?;
    client.create_database(db2_name).await?;

    // Switch to db1 and create a node
    client.switch_database(db1_name).await?;
    let mut params = HashMap::new();
    params.insert("name".to_string(), Value::String("DB1 Node".to_string()));
    let result = client
        .execute_cypher("CREATE (n:TestNode {name: $name}) RETURN n", Some(params))
        .await?;
    assert_eq!(result.rows.len(), 1);

    // Verify node exists in db1
    let count_result = client
        .execute_cypher("MATCH (n:TestNode) RETURN count(n) AS count", None)
        .await?;
    // Row is an array: [count_value]
    let count = count_result.rows[0].as_array().unwrap()[0]
        .as_i64()
        .unwrap();
    assert_eq!(count, 1);

    // Switch to db2
    client.switch_database(db2_name).await?;

    // Verify node does NOT exist in db2 (isolation)
    let count_result = client
        .execute_cypher("MATCH (n:TestNode) RETURN count(n) AS count", None)
        .await?;
    let count = count_result.rows[0].as_array().unwrap()[0]
        .as_i64()
        .unwrap();
    assert_eq!(count, 0);

    // Create a different node in db2
    let mut params = HashMap::new();
    params.insert("name".to_string(), Value::String("DB2 Node".to_string()));
    let result = client
        .execute_cypher("CREATE (n:TestNode {name: $name}) RETURN n", Some(params))
        .await?;
    assert_eq!(result.rows.len(), 1);

    // Verify only one node in db2
    let count_result = client
        .execute_cypher("MATCH (n:TestNode) RETURN count(n) AS count", None)
        .await?;
    let count = count_result.rows[0].as_array().unwrap()[0]
        .as_i64()
        .unwrap();
    assert_eq!(count, 1);

    // Switch back to db1
    client.switch_database(db1_name).await?;

    // Verify still only one node in db1
    let count_result = client
        .execute_cypher("MATCH (n:TestNode) RETURN count(n) AS count", None)
        .await?;
    let count = count_result.rows[0].as_array().unwrap()[0]
        .as_i64()
        .unwrap();
    assert_eq!(count, 1);

    // Clean up
    let databases = client.list_databases().await?;
    client.switch_database(&databases.default_database).await?;
    client.drop_database(db1_name).await?;
    client.drop_database(db2_name).await?;

    Ok(())
}

#[tokio::test]
async fn test_client_with_database_parameter() -> Result<(), NexusError> {
    // Create a test database first
    let setup_client = NexusClient::new("http://localhost:15474")?;

    let db_name = "test_param_db";
    setup_client.create_database(db_name).await?;

    // Create a client and switch to the specific database
    let client = NexusClient::new("http://localhost:15474")?;
    client.switch_database(db_name).await?;

    // Verify we're connected to the right database
    let current_db = client.get_current_database().await?;
    assert_eq!(current_db, db_name);

    // Clean up
    let databases = setup_client.list_databases().await?;
    setup_client
        .switch_database(&databases.default_database)
        .await?;
    setup_client.drop_database(db_name).await?;

    Ok(())
}

#[tokio::test]
async fn test_cannot_drop_current_database() -> Result<(), NexusError> {
    let client = NexusClient::new("http://localhost:15474")?;

    let db_name = "test_no_drop_db";

    // Create a test database
    client.create_database(db_name).await?;

    // Switch to the database
    client.switch_database(db_name).await?;

    // Try to drop it while it's active - should fail
    let drop_result: Result<_, NexusError> = client.drop_database(db_name).await;
    assert!(drop_result.is_err());

    // Switch to a different database
    let databases = client.list_databases().await?;
    client.switch_database(&databases.default_database).await?;

    // Now we should be able to drop it
    let drop_result = client.drop_database(db_name).await?;
    assert!(drop_result.success);

    Ok(())
}

#[tokio::test]
async fn test_cannot_drop_default_database() -> Result<(), NexusError> {
    let client = NexusClient::new("http://localhost:15474")?;

    // Get default database
    let databases = client.list_databases().await?;
    let default_db = &databases.default_database;

    // Try to drop it - should fail
    let drop_result: Result<_, NexusError> = client.drop_database(default_db).await;
    assert!(drop_result.is_err());

    Ok(())
}
